### PR TITLE
RFC: Console Application improvements (arguments bag, command validation, aliases, argument injectors, mistyped commands suggestions, themes)

### PR DESCRIPTION
**Pull request by przemyslaw-przylucki**
This PR aims to enhance the capabilities of the CLI layer 🎉 

image

Related to #225, it has a lot of rough edges - I've rushed a proof of concept in a few hours. So it's more like an RFC :)

Introduced changes:
- Global ArgumentBag that holds arguments passed to commands
- ConsoleCommands now get validated - annotations may be added to validate the argument (we may think about adding type-rules to, either to use manually or for validator to infer from reflection)
- Adds predefined flags (eventually arguments) - that are easily used in user-land code (i.e by passing isDangerous to ConsoleCommand attribute)
- These flags kinda act like a middleware for commands - maybe it's worth to add a "pipeline" support class?
- The "listing command" will also discover injected arguments
- Added "--help (-h)", "--force (-f)" "--silent (-s)" "--no-interaction (-n)" flags to all commands
- Added ability to alias commands

image

Left to do:
- tests 🥶

### CODE REVIEW

File: src/Console/ConsoleCommand.php [hunk 1]
```diff
@@ -55,9 +66,68 @@ public function __unserialize(array $data): void
...
+        return new CommandArguments(
+            arguments: $arguments,
+            injectedArguments: $injected,
+        );
+    }
+
+    /**
+     * @return InjectedArgument[]
+     */
+    protected function flagList(): array
```
Comment (przemyslaw-przylucki) [inline src/Console/ConsoleCommand.php:121]: That should probably be discovered by the framework instead

**Comment by przemyslaw-przylucki** [comment]
Also, would you be open to making the ConsoleCommand attribute target the entire class and not a handler?

The negative is that we would have to infer the default argument order based on the properties present in the class, but I think it'd be easier to add additional information to command parameters.

I've started exploring what's out there after building this and I mostly love the api this package provides - https://github.com/thettler/laravel-console-toolkit

**Comment by aidan-casey** [comment]
Hey, @przemyslaw-przylucki! I haven't had the time to review everything here, but I think there is some stuff here that we should definitely think about. I hope to have some more helpful comments here soon. 🙂 

I'm going to defer to @brendt on the question about class-based commands. My guess is that will be a no, however.

**Comment by przemyslaw-przylucki** [comment]
I didn't intend this to grow in scope so much but I went coo-coo and couldn't stop polishing this - still a lot of things that can be improved but I like how it stands from features perspective


Also, pretty rad UI if you ask me:

image

**Comment by przemyslaw-przylucki** [comment]
I've pushed some additional code, they contains:
- console output builder
- suggested/similar commands, if user mistypes command and levenstein distance is <= 3
- ConsoleTheme - I think it'd be nice if we had a layer of abstraction on top of ConsoleOutput that can format strings, this would be nice for defining a style guide for some package authors or internal commands

image

**Comment by brendt** [comment]
Ok this is an impressive PR 💪

It'll take some time to read through it! Please hold :)

File: src/Support/Reflection/Attributes.php [hunk 2]
```diff
@@ -7,12 +7,13 @@
...
 use ReflectionClass;
 use ReflectionMethod;
+use ReflectionParameter;
 use ReflectionProperty;
 
 /** @template T */
 final class Attributes
 {
-    private ReflectionClass|ReflectionMethod|ReflectionProperty $reflector;
+    private ReflectionClass|ReflectionMethod|ReflectionProperty|ReflectionParameter $reflector;
```
Comment (brendt) [inline src/Support/Reflection/Attributes.php:16]: Outside the scope of this PR, but we'll need to clean this part up :p

File: src/Validation/NestedValidator.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+namespace Tempest\Validation;
+
+use ReflectionClass;
+use ReflectionProperty;
+use function Tempest\attribute;
+use Tempest\Console\Argument;
+use Tempest\Validation\Exceptions\ValidationException;
+
+// todo: just a proof of concept.
```
Comment (brendt) [inline src/Validation/NestedValidator.php:13]: Can you explain the concept and why it's needed?

File: src/Console/Commands/MigrateCommand.php [hunk 4]
```diff
@@ -19,23 +18,21 @@ final class MigrateCommand
     public function __construct(
         private readonly Console $console,
         private readonly MigrationManager $migrationManager,
-        private readonly AppConfig $config,
     ) {
     }
 
     #[ConsoleCommand(
         name: 'migrate:up',
+        aliases: ['haha'],
```
Comment (brendt) [inline src/Console/Commands/MigrateCommand.php:26]: This will need to change :)

File: src/Console/Commands/MigrateCommand.php [hunk 5]
```diff
@@ -19,23 +18,21 @@ final class MigrateCommand
...
         description: 'Run all new migrations',
+        isDangerous: true,
+        help: [
```
Comment (brendt) [inline src/Console/Commands/MigrateCommand.php:29]: Why does help text need to be an array? Wouldn't a string suffice?

File: src/Console/ConsoleCommand.php [hunk 6]
```diff
@@ -6,16 +6,28 @@
...
 use Attribute;
 use ReflectionMethod;
+use Tempest\Console\Arguments\ForceArgument;
+use Tempest\Console\Arguments\HelpArgument;
+use Tempest\Console\Arguments\NoInteractionArgument;
+use Tempest\Console\Arguments\QuietArgument;
 
 #[Attribute]
-final class ConsoleCommand
+final class ConsoleCommand implements HasHelpLines
```
[the remark this answers is not in the document]
Comment (brendt) [reply src/Console/ConsoleCommand.php:15]: To add: why not a string instead of an array?

File: src/Console/HandlesConsoleInput.php [hunk 7]
```diff
@@ -27,22 +29,16 @@ public function ask(
...
-                    ),
-                ),
-            )
-                . '] ';
-        }
-
-        $this->output->write($questionString);
+        $message = ConsoleOutputBuilder::new(" ")
+            ->error(" $question ")
+            ->when(! ! $options, function (ConsoleOutputBuilder $builder) use ($options, $default) {
```
Comment (brendt) [inline src/Console/HandlesConsoleInput.php:34]: I guess the double `!!` is to convert it to booleans? Why not `$options !== null` ?

File: src/Console/HandlesConsoleInput.php [hunk 8]
```diff
@@ -27,22 +29,16 @@ public function ask(
...
-                        fn (string $option) => $option === $default ? strtoupper($option) : $option,
-                        $options,
-                    ),
-                ),
-            )
-                . '] ';
-        }
-
-        $this->output->write($questionString);
+        $message = ConsoleOutputBuilder::new(" ")
```
Comment (brendt) [inline src/Console/HandlesConsoleInput.php:32]: I like this concept!

File: src/Console/HandlesConsoleInput.php [hunk 9]
```diff
@@ -27,22 +29,16 @@ public function ask(
...
+            ->error(" $question ")
+            ->when(! ! $options, function (ConsoleOutputBuilder $builder) use ($options, $default) {
+                $builder->formatted("[" . ConsoleStyle::FG_BLUE(implode(', ', $options)) . "]")
+                    ->formatted($default ? ConsoleStyle::FG_GRAY(" (default: $default) ") : " ");
+            });
+
+        $this->output->write(
```
Comment (brendt) [inline src/Console/HandlesConsoleInput.php:39]: Maybe add a convenience method on the builder class? `write(to: $output)`

File: src/Console/InjectedArgument.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+     * @throws ExitException
+     */
+    public function handle(ConsoleCommand $command): void
+    {
+
+    }
+
+    public function shouldInject(): bool
+    {
+        return ! ! $this->value;
```
Comment (brendt) [inline src/Console/InjectedArgument.php:21]: Double `!!`

File: src/Console/InjectedArgument.php [hunk 11]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console;
+
+use ReflectionParameter;
+
+readonly class InjectedArgument extends Argument
```
Comment (brendt) [inline src/Console/InjectedArgument.php:9]: Deep inheritance chain should be refactored

File: src/Console/NullArgumentBag.php [hunk 12]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console;
+
+use InvalidArgumentException;
+
+final class NullArgumentBag implements ArgumentBag
```
Comment (brendt) [inline src/Console/NullArgumentBag.php:9]: Is this for testing? We should probably move it to `Tempest\Testing` if that's the case

File: src/Console/Styling/ConsoleOutputThemeInitializer.php [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+declare(strict_types=1);
+
+namespace Tempest\Console\Styling;
+
+use Tempest\Container\Container;
+use Tempest\Container\Initializer;
+use Tempest\Container\Singleton;
+
+#[Singleton]
+final readonly class ConsoleOutputThemeInitializer implements Initializer
```
Comment (brendt) [inline src/Console/Styling/ConsoleOutputThemeInitializer.php:12]: Should a theme really be registered in the container? I figure it's just a config value somewhere that's used when constructing the console?

File: src/Console/Styling/OutputType.php [hunk 14]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console\Styling;
+
+enum OutputType
```
Comment (brendt) [inline src/Console/Styling/OutputType.php:7]: `ConsoleOutputType` is probably clearer

File: src/Console/Styling/OutputLine.php [hunk 15]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console\Styling;
+
+final class OutputLine
```
Comment (brendt) [inline src/Console/Styling/OutputLine.php:7]: `ConsoleOutputLine` is probably clearer

File: src/Console/Styling/RenderCommandNotFound.php [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public function __invoke(string $command): string
+    {
+        $similarCommands = $this->getSimilarCommands($command);
+
+        return ConsoleOutputBuilder::new()
+            ->error("Command `{$command}` not found")
+            ->blank()
+            ->when(! ! $similarCommands, function (ConsoleOutputBuilder $response) use ($similarCommands) {
```
Comment (brendt) [inline src/Console/Styling/RenderCommandNotFound.php:27]: `!!`

File: src/Console/Styling/RenderCommandNotFound.php [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+                $response->warning('Did you mean one of these commands?')
+                    ->blank();
+
```
Comment (brendt) [inline src/Console/Styling/RenderCommandNotFound.php:30]: Something we could add along the way is to allow the user to pick the command automatically from the error output without having to restart the console session. But doesn't need to be added now.

File: src/Console/Styling/RenderConsoleCommandOverview.php [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        private RenderConsoleCommand $renderrer,
+    ) {
+    }
+
+    public function __invoke(): string
+    {
+        $commands = $this->getCommands();
+
+        $response = ConsoleOutputBuilder::new()
+            ->brand(' Tempest Console ')
```
Comment (brendt) [inline src/Console/Styling/RenderConsoleCommandOverview.php:26]: The surrounding spaces should probably be added within the `brand` method?

File: src/Console/Styling/RenderConsoleCommandOverview.php [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        private ConsoleConfig $consoleConfig,
+        private RenderConsoleCommand $renderrer,
+    ) {
+    }
+
+    public function __invoke(): string
+    {
+        $commands = $this->getCommands();
+
+        $response = ConsoleOutputBuilder::new()
```
Comment (brendt) [inline src/Console/Styling/RenderConsoleCommandOverview.php:25]: Just once again: I really love this output builder!

**Comment by przemyslaw-przylucki** [comment]
Thanks for all the feedback - I agree with everything from the design perspective, I prefer to go wild and then refactor =)

I'd love to get this PR finalized - I think breaking this into multiple smaller PRs def makes sense - wanted to cram this bad boy with a lot of convenience features. I think there's a lot of potential in the console layer of the project and I'd love to continue working on it if it's alright with y'all!

I'll respond to individual comments in the evening :D

**Comment by przemyslaw-przylucki** [comment]
> * Are you committed to finishing this PR? It'll probably take a while before we can merge it, as it's pretty huge. I'm totally fine if you don't have the capacity to do so, but I'd just like some clarity up front, so that I know how what's expected from my side.

One of the things that this PR needs is the type inference and changes in validation that were added in #229, could you let me know if that's something you'd want to go forward with?

**Comment by brendt** [comment]
> One of the things that this PR needs is the type inference and changes in validation that were added in https://github.com/tempestphp/tempest-framework/pull/229, could you let me know if that's something you'd want to go forward with?

I still need to go through that one, probably won't happen today though. What I can tell from a quick scan is that we definitely need this, we might just need to finetune the details.

File: src/Console/Commands/MigrateCommand.php (hunk 5 above)
Comment (przemyslaw-przylucki) [reply src/Console/Commands/MigrateCommand.php:29]: I'd go with string | string[]. If user wants to create a longer description, or add some white space, this allows you for an easier control while still being convenient

File: src/Console/Styling/ConsoleOutputThemeInitializer.php (hunk 13 above)
Comment (przemyslaw-przylucki) [reply src/Console/Styling/ConsoleOutputThemeInitializer.php:12]: I don't remember why I went with that design.

I think you are right - the ConsoleApplication should probably pull the base theme from the ConsoleConfig. 

When it comes to user-defined themes:

We could add a property to ConsoleCommand there to overwrite the default theme. 

If we want to go crazy and drop final from ConsoleCommand, both 1st and 3rd party packages or organizations could create their own styling, which is pretty neat from package author "branding" perspective.

This way they could create a simple wrapper around ConsoleCommand and use their own brand i.e PlatypusCorpCommand

File: src/Console/ArgumentBagInitializer.php [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+use Tempest\Container\Singleton;
+
+#[Singleton]
+final readonly class ArgumentBagInitializer implements Initializer
+{
+    public function initialize(Container $container): ArgumentBag
+    {
+        $arguments = $_SERVER['argv'];
+
+        return match (! ! $arguments) {
```
[the remark this answers is not in the document]
Comment (przemyslaw-przylucki) [reply src/Console/ArgumentBagInitializer.php:18]: I love !! as a quick boolean cast - works in most languages

File: src/Support/BaseBuilder.php [hunk 21]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Support;
+
+class BaseBuilder
```
[the remark this answers is not in the document]
Comment (przemyslaw-przylucki) [reply src/Support/BaseBuilder.php:7]: I'll go with a for now ;p

File: src/Console/ArgumentBagInitializer.php (hunk 20 above)
[the remark this answers is not in the document]
Comment (brendt) [reply src/Console/ArgumentBagInitializer.php:18]: Maybe it needs some getting used to 🤔

File: src/Support/BaseBuilder.php (hunk 21 above)
[the remark this answers is not in the document]
Comment (brendt) [reply src/Support/BaseBuilder.php:7]: Yeah I like the trait approach better 😁 thanks!

File: src/Console/Commands/MigrateCommand.php (hunk 5 above)
Comment (brendt) [reply src/Console/Commands/MigrateCommand.php:29]: Wouldn't a heredoc be sufficient in that case? I'm just trying to imagine when an array would really be helpful, and I can't really come up with a good example.

File: src/Console/Styling/ConsoleOutputThemeInitializer.php (hunk 13 above)
Comment (brendt) [reply src/Console/Styling/ConsoleOutputThemeInitializer.php:12]: > If we want to go crazy and drop final from ConsoleCommand, both 1st and 3rd party packages or organizations could create their own styling, which is pretty neat from package author "branding" perspective.

We don't want to go crazy now, do we? 😛

Sidenote: let me actually share some background info while we're at it: I'm considering to extract the console code into a separate package that people can use as standalone console replacement. I figure it's about time that `symfony/console` gets some proper competition 😛

That won't happen before we merge all these features though, and we shouldn't worry too much about it for now. I'll take care of extracting stuff all at once. I might need to make some changes down the line for things to work in a standalone context.

/sidenote

So regarding personalization, I think registering the output builder within the container is the way to go, and have the initializer take care of setting up branding. To me, that feels like the approach that's most in line with the "Tempest way"

**Pull request closed without merging**